Refresh.GameServer/Middlewares/DigestMiddleware.cs

```diff
@@ -1,114 +1,158 @@
             return;
         }
 
-        short? exeVersion = null;
-        short? dataVersion = null;
-        if (short.TryParse(context.RequestHeaders["X-Exe-V"], out short exeVer))
+        foreach (string digest in config.Sha1DigestKeys)
         {
-            exeVersion = exeVer;
+            string calculatedClientDigest = CalculateDigest(digest, route, context.InputStream, auth, pspVersionInfo, isUpload, false);
+
+            // If the calculated client digest is invalid, then this isn't the digest the game is using, so check the next one
+            if (calculatedClientDigest != clientDigest) 
+                continue;
+
+            SetDigestResponse(context, CalculateDigest(digest, route, context.ResponseStream, auth, null, isUpload, false));
+            
+            if(token != null)
+                gameDatabase.SetTokenDigestInfo(token, digest, false);
+            
+            return;
         }
-        if (short.TryParse(context.RequestHeaders["X-Data-V"], out short dataVer))
+
+        foreach (string digest in config.HmacDigestKeys)
         {
-            dataVersion = dataVer;
+            string calculatedClientDigest = CalculateDigest(digest, route, context.InputStream, auth, pspVersionInfo, isUpload, true);
+
+            // If the calculated client digest is invalid, then this isn't the digest the game is using, so check the next one
+            if (calculatedClientDigest != clientDigest) 
+                continue;
+            
+            SetDigestResponse(context, CalculateDigest(digest, route, context.ResponseStream, auth, null, isUpload, true));
+            
+            if(token != null)
+                gameDatabase.SetTokenDigestInfo(token, digest, true);
+            
+            return; 
         }
 
-        this.VerifyDigestRequest(context, exeVersion, dataVersion);
-        Debug.Assert(context.InputStream.Position == 0); // should be at position 0 before we pass down the pipeline
+        // If we were unable to find any digests, just use the first one specified as a backup
+        // TODO: once we have PS4 support, check if the token is a PS4 token
+        bool isPs4 = context.RequestHeaders["User-Agent"] == "MM CHTTPClient LBP3 01.26";
+        string firstDigest = isPs4 ? config.HmacDigestKeys[0] : config.Sha1DigestKeys[0];
         
-        next();
-
-        // should be at position 0 before we try to set digest
-        context.ResponseStream.Seek(0, SeekOrigin.Begin);
-        this.SetDigestResponse(context);
+        SetDigestResponse(context, CalculateDigest(firstDigest, route, context.ResponseStream, auth, null, isUpload, isPs4));
+        
+        if(token != null)
+            gameDatabase.SetTokenDigestInfo(token, firstDigest, isPs4);
     }
+
+    private static void SetDigestResponse(ListenerContext context, string calculatedDigest)
+        => context.ResponseHeaders["X-Digest-A"] = calculatedDigest;
```

Comment:
probably doesn't need to be a function at all?

Reply:
doesn't directly need to be, but it felt ugly to read `context.ResponseHeaders["X-Digest-A"] =` typed out multiple times, so i consolidated it to be easier to read at a glance, i can remove the function if wanted

Reply:
Was mostly just wondering, feel free to keep